util-images/probes/pkg/dnspropagation/probe.go

```diff
@@ -0,0 +1,234 @@
+limitations under the License.
+*/
+
+package dnspropagation
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"math"
+	"math/rand"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+var (
+	statefulSet     = flag.String("dns-propagation-probe-stateful-set", "", "Name of the statefulSet workload")
+	service         = flag.String("dns-propagation-probe-service", "", "Name of the headless service that exposes the statefulSet resides")
+	namespacePrefix = flag.String("dns-propagation-probe-namespace-prefix", "test-", "Prefix of the namespace where the statefulSet resides")
+	clusterDomain   = flag.String("dns-propagation-probe-cluster-domain", "cluster", "Name of cluster domain where the statefulSet resides")
+	suffix          = flag.String("dns-propagation-probe-suffix", "local", "DNS label suffix")
+	interval        = flag.Duration("dns-propagation-probe-interval", 100*time.Millisecond, "Interval between DNS lookups")
+	podCount        = flag.Int("dns-propagation-probe-pod-count", 0, "Number of pods in the statefulSet")
+	sampleCount     = flag.Int("dns-propagation-probe-sample-count", 0, "Number of pods to test dns propagation against in the statefulSet, defaults to min(100, Ceil(SQRT(podCount))")
+)
+
+type DNSPodPropagationResult struct {
+	podName  string
+	duration time.Duration
+}
+
+// Run periodically does DNS lookups.
+// Lookup's interval and URL are configurable via flags.
+func Run() {
+	if *statefulSet == "" {
+		klog.Fatal("--dns-propagation-probe-stateful-set has not been set")
+	}
+	if *service == "" {
+		klog.Fatal("--dns-propagation-probe-service-set has not been set")
+	}
+	if *podCount <= 0 {
+		klog.Fatal("--dns-propagation-probe-pod-count has not been set or is not a positive number")
+	}
+	if *sampleCount <= 0 {
+		f := int(math.Ceil(math.Sqrt(float64(*podCount))))
+		f = int(math.Min(float64(f), 100))
+		sampleCount = &f
+	}
+	// creates the in-cluster config
+	kubeConfig, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatalf("Can not build inClusterConfig, error:%v", err)
+	}
+	// creates the inCluster kube client
+	clientset, err := kubernetes.NewForConfig(kubeConfig)
+
+	if err != nil {
+		klog.Fatalf("Failed to build kubeClient, error:%v", err)
+	}
+	//TODO deprecated as of Go 1.20. To remove when go version gets bumped
+	rand.Seed(time.Now().UnixNano())
+	runProbe(clientset)
+	for {
+		klog.V(2).Infof("dns propagation probe complete, waiting until the test finishes...")
+		time.Sleep(10 * time.Second)
+	}
+}
+func runProbe(client kubernetes.Interface) {
+	klog.V(1).Infof("DNS propagation probe started")
+	var namespace string
+	for {
+		klog.V(2).Infof("Get namespace from client based on prefix %s", *namespacePrefix)
+		fullNamespace, err := getNamespacesFromClient(client, *namespacePrefix)
```

Comment:
This might work randomly if you have multiple namespaces matching the prefix.

If you intend to execute this container image from clusterloader2/pkg/measurement/common/probes/probes.go, then, there, you will have access to `*measurement.Config`: https://github.com/kubernetes/perf-tests/blob/32357a4c9514cca8cd61b4c327547adc686b68a3/clusterloader2/pkg/measurement/interface.go#L29-L49 which has the `ClusterFramework` field: https://github.com/kubernetes/perf-tests/blob/32357a4c9514cca8cd61b4c327547adc686b68a3/clusterloader2/pkg/framework/framework.go#L44-L56 which has two fields that could be of interest to you:

- `automanagedNamespacePrefix`
- `automanagedNamespaces`

IMHO we can get simpler (overall) approach working here:
- Pass the index of the namespace of our interest to the measurement through `Params`,
- Reconstruct the namespace name using that index and aforementioned `automanagedNamespacePrefix` field,
- Use the reconstructed namespace name when filling out the template for the prober's deployment.

Thoughts?

Reply:
Done